sql-plugin/src/main/scala/com/nvidia/spark/rapids/aggregate.scala

```diff
@@ -70,437 +75,362 @@ object AggregateUtils {
+          }
+        }
+      }
 
-  override def verboseStringWithOperatorId(): String = {
-    s"""
-       |$formattedNodeName
-       |${ExplainUtils.generateFieldString("Input", child.output)}
-       |${ExplainUtils.generateFieldString("Keys", groupingExpressions)}
-       |${ExplainUtils.generateFieldString("Functions", aggregateExpressions)}
-       |${ExplainUtils.generateFieldString("Aggregate Attributes", aggregateAttributes)}
-       |${ExplainUtils.generateFieldString("Results", resultExpressions)}
-       |""".stripMargin
-  }
+      val mergedBatch = if (batchesToConcat.length > 1) {
+        wasBatchMerged = true
+        val batch = concatenateAndMerge(batchesToConcat)
+        batch.allowSpilling()
+        batch
+      } else {
+        // Unable to find a neighboring buffer to produce a valid merge in this pass,
+        // so simply put this buffer back on the queue for other passes.
+        batchesToConcat.remove(0)
+      }
 
-  case class BoundExpressionsModeAggregates(boundInputReferences: Seq[GpuExpression] ,
-    boundFinalProjections: Option[scala.Seq[GpuExpression]],
-    boundResultReferences: scala.Seq[Expression] ,
-    aggModeCudfAggregates: scala.Seq[(AggregateMode, scala.Seq[CudfAggregate])])
-  // This handles GPU hash aggregation without spilling.
-  //
-  // The CPU version of this is (assume no fallback to sort-based aggregation)
-  //  Re: TungstenAggregationIterator.scala, and AggregationIterator.scala
-  //
-  // 1) Obtaining an input row and finding a buffer (by hash of the grouping key) where
-  //    to aggregate on.
-  // 2) Once it has a buffer, it calls processRow on it with the incoming row.
-  // 3) This will in turn update the buffer, for each row received, agg function by agg function
-  // 4) In the happy case, we never spill, and an iterator (from the HashMap) is produced s.t.
-  //    downstream nodes can access the aggregated and projected results.
-  // 5) Spill case (not handled in gpu case)
-  //    a) we create an external sorter [[UnsafeKVExternalSorter]] from the hash map (spilling)
-  //       this leaves room for more aggregations to happen. The external sorter first sorts, and
-  //       then stores the results to disk, and last it clears the in-memory hash map.
-  //    b) we merge external sorters as we spill the map further.
-  //    c) after the hash agg is done with its incoming rows, it will switch to sort-based
-  //       aggregation, because it detected a spill
-  //    d) sort based aggregation is then the mode forward, and not covered in this.
-  override def doExecuteColumnar(): RDD[ColumnarBatch] = {
-    val numOutputRows = gpuLongMetric(NUM_OUTPUT_ROWS)
-    val numOutputBatches = gpuLongMetric(NUM_OUTPUT_BATCHES)
-    val totalTime = gpuLongMetric(TOTAL_TIME)
-    val computeAggTime = gpuLongMetric(AGG_TIME)
-    val concatTime = gpuLongMetric(CONCAT_TIME)
-    val rdd = child.executeColumnar()
+      aggregatedBatches.addLast(mergedBatch)
```

Comment:
nit: can we add a comment here about how we are doing a single merge pass through the queue and even though we appending the results to the end of the queue, we prevent them from taking part in the merge step again.